### ISSUE TITLE
torchcodec decode extremely slow at large mkv file(42x slower than mp4)

### ISSUE THREAD

**Issue by MilkClouds** [meta-pytorch/torchcodec#1223]
### 🐛 Describe the bug

On following code, decoding frames on mkv is 42x slower than mp4. It occurs when **start time>>0.** (work normally at start_time=0)

```
Benchmarking 5.0s window @ 10.0Hz
Start: 250.0s
MKV: /mnt/testfs/example-mcaps/woozy-ruby-ostrich-f153ac423f61-20220224-172750.mkv
MP4: /mnt/testfs/example-mcaps/woozy-ruby-ostrich-f153ac423f61-20220224-172750.mp4
------------------------------------------------------------
MKV: 5.8444s (50 frames)
MP4: 0.1376s (50 frames)
------------------------------------------------------------
MKV/MP4 ratio: 42.47x
```

Following is code I used for experiment. Full video is 300s length.

```py
#!/usr/bin/env python3
"""Benchmark script comparing torchcodec read times for MKV vs MP4 at 10Hz, 5-second window."""

import time
from pathlib import Path

from torchcodec.decoders import VideoDecoder

# Video paths
BASE_NAME = "woozy-ruby-ostrich-f153ac423f61-20220224-172750"
VIDEO_DIR = Path("/mnt/testfs/example-mcaps")
MKV_PATH = VIDEO_DIR / f"{BASE_NAME}.mkv"
MP4_PATH = VIDEO_DIR / f"{BASE_NAME}.mp4"

# Sampling parameters
FPS = 10.0  # 10Hz
WINDOW_SECONDS = 5.0  # 5-second window
START_SECONDS = 250.0  # start at 250s


def benchmark_read(video_path: Path, start_seconds: float = START_SECONDS) -> tuple[float, int]:
    """Read a 5-second window at 10Hz and return (elapsed_time, num_frames)."""
    decoder = VideoDecoder(video_path, seek_mode="approximate")

    # Generate timestamps at 10Hz for 5 seconds: 0.0, 0.1, 0.2, ..., 4.9
    num_frames = int(WINDOW_SECONDS * FPS)
    timestamps = [start_seconds + i / FPS for i in range(num_frames)]

    t0 = time.perf_counter()
    frames = decoder.get_frames_played_at(seconds=timestamps)
    elapsed = time.perf_counter() - t0

    return elapsed, frames.data.shape[0]


def main():
    print(f"Benchmarking {WINDOW_SECONDS}s window @ {FPS}Hz")
    print(f"Start: {START_SECONDS}s")
    print(f"MKV: {MKV_PATH}")
    print(f"MP4: {MP4_PATH}")
    print("-" * 60)

    # Warm-up (optional, helps with cache effects)
    for path in [MKV_PATH, MP4_PATH]:
        if path.exists():
            benchmark_read(path)

    # Benchmark MKV
    if MKV_PATH.exists():
        mkv_time, mkv_frames = benchmark_read(MKV_PATH)
        print(f"MKV: {mkv_time:.4f}s ({mkv_frames} frames)")
    else:
        print(f"MKV: NOT FOUND")

    # Benchmark MP4
    if MP4_PATH.exists():
        mp4_time, mp4_frames = benchmark_read(MP4_PATH)
        print(f"MP4: {mp4_time:.4f}s ({mp4_frames} frames)")
    else:
        print(f"MP4: NOT FOUND")

    # Compare
    if MKV_PATH.exists() and MP4_PATH.exists():
        print("-" * 60)
        ratio = mkv_time / mp4_time
        print(f"MKV/MP4 ratio: {ratio:.2f}x")


if __name__ == "__main__":
    main()

```

### Versions

```
Collecting environment information...
PyTorch version: 2.10.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: Could not collect
CMake version: version 4.1.2
Libc version: glibc-2.35

Python version: 3.11.14 | packaged by conda-forge | (main, Oct 22 2025, 22:46:25) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1091-nvidia-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 580.95.05
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  224
On-line CPU(s) list:                     0-223
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Platinum 8480C
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      56
Socket(s):                               2
Stepping:                                8
CPU max MHz:                             3800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4000.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               5.3 MiB (112 instances)
L1i cache:                               3.5 MiB (112 instances)
L2 cache:                                224 MiB (112 instances)
L3 cache:                                210 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-55,112-167
NUMA node1 CPU(s):                       56-111,168-223
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.8
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.2
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] torch==2.10.0
[pip3] torchaudio==2.8.0
[pip3] torchcodec==0.10.0
[pip3] torchvision==0.25.0
[pip3] triton==3.6.0
[conda] libfaiss                  1.12.0          py3.11_hf65b397_0_cpu    pytorch
[conda] libopenvino-pytorch-frontend 2025.3.0             hecca717_0    conda-forge
[conda] mkl                       2023.0.0         h84fe81f_26648    conda-forge
[conda] msgpack-numpy             0.4.8                    pypi_0    pypi
[conda] numpy                     2.4.2                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.4.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.90                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.93                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.90                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.10.2.21                pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.83                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.90                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.3.90                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.8.93                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.7.1                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.27.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.93                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.90                  pypi_0    pypi
[conda] tbb                       2021.13.0            hb60516a_3    conda-forge
[conda] torch                     2.10.0                   pypi_0    pypi
[conda] torchaudio                2.8.0                    pypi_0    pypi
[conda] torchcodec                0.10.0                   pypi_0    pypi
[conda] torchvision               0.25.0                   pypi_0    pypi
[conda] triton                    3.6.0                    pypi_0    pypi
```

**Comment by MilkClouds** [comment]
Both mkv/mp4 file have same GOP structure.
```
Analyzing frames in /mnt/testfs/example-mcaps/woozy-ruby-ostrich-f153ac423f61-20220224-172750.mkv...

Video Information:
  Codec: h264
  Resolution: 640x360
  Frame Rate: 20.00 fps
  Duration: 0.00 seconds

Frame Type Distribution:
  H.26[45] User Data Unregistered SEI message-frames: 1 (0.0%)
  P-frames: 5800 (96.7%)
  I-frames: 200 (3.3%)

I-frame Interval Analysis:
  Total I-frames: 200
  I-frame intervals (seconds):
    Minimum: 1.50
    Maximum: 1.50
    Average: 1.50
    Median:  1.50

  Consistent GOP size detected: 1.50 seconds

Frame Pattern (first 200 frames):
H.26[45] User Data Unregistered SEI messagePPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP
```

```
Analyzing frames in /mnt/testfs/example-mcaps/woozy-ruby-ostrich-f153ac423f61-20220224-172750.mp4...

Video Information:
  Codec: h264
  Resolution: 640x360
  Frame Rate: 20.00 fps
  Duration: 300.05 seconds

Frame Type Distribution:
  H.26[45] User Data Unregistered SEI message-frames: 1 (0.0%)
  P-frames: 5800 (96.7%)
  I-frames: 200 (3.3%)

I-frame Interval Analysis:
  Total I-frames: 200
  I-frame intervals (seconds):
    Minimum: 1.50
    Maximum: 1.50
    Average: 1.50
    Median:  1.50

  Consistent GOP size detected: 1.50 seconds

Frame Pattern (first 200 frames):
H.26[45] User Data Unregistered SEI messagePPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP PPPPPPPPPP IPPPPPPPPP PPPPPPPPPP
```

**Comment by MilkClouds** [comment]
Also, not only speed, but also I/O bandwidth(measured from bytes read from temporary filesystem. video being read is saved on temporary, separate filesystem on benchmarking) is hugely heavier for mkv. p.s. Since the code used for benchmarking is closely coupled with our training code it's difficult to share code for this figure.

<img width="1778" height="301" alt="Image" src="https://github.com/user-attachments/assets/002553b8-78ec-4ab3-84ee-33345ba1d937" />

time: in seconds.
read: bytes read, in MB.
x-axis is `iter(dataset)`'s index, which means same step for mkv/mp4 = same frame count decoded on mkv/mp4.

**Comment by Dan-Flores** [comment]
Thanks for the detailed issue @MilkClouds

I noticed the `mkv` file is missing duration metadata. This might cause `VideoDecoder` to do additional seeks when using `seek_mode=approximate`. 

Could you try using `seek_mode=exact`? This adds the upfront cost of scanning the video to build an index and gather correct metadata for the video, but should result in faster decoding than the performance you are seeing here.

If you are able to share the video file with this issue, that would be helpful in debugging this! Its possible the video is missing additional metadata that is causing this slowness.

**Comment by MilkClouds** [comment]
@Dan-Flores 

Thanks for the reply! I tried `seek_mode=exact`. The decoding speed has normalized, but it uses an enormous amount of disk I/O. (Because of this issue, I absolutely intend to use the approximate setting.) On experiments for following figure, I'm testing with 64 5min video for minecraft, so `seek_mode=exact` needs to scan 64*5=320min.

<img width="1755" height="307" alt="Image" src="https://github.com/user-attachments/assets/867f24a5-e7f8-4e33-a33a-c931478cb471" />

As long as I know, mkv does have duration info, but it has only `container.duration`, not `stream.duration` (but not I'm sure). I can share video pair(mkv/mp4): https://drive.google.com/file/d/1f46hpsyEhMDwsbEY712CdTeLNw-MlPjT/view?usp=drive_link . I would be greatly helpful if you can find the reason why mkv format is troublesome.

**Comment by Dan-Flores** [comment]
Thanks for the reproducing example @MilkClouds, we were able to resolve this bug and will include the fix in the next release!
